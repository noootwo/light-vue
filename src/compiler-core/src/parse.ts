import { NodeTypes } from "./ast";

const Delimiter = {
  interpolationOpen: "{{",
  interpolationClose: "}}",
};

export function baseParse(content: string) {
  const context = createParseContext(content);

  return createRoot(parseChildren(context));
}

function createRoot(children) {
  return {
    children,
  };
}

function parseChildren(context) {
  const nodes: any = [];

  if (context.source.startsWith(Delimiter.interpolationOpen)) {
    const node = parseInterpolation(context);

    nodes.push(node);
  }

  return nodes;
}

function parseInterpolation(context) {
  const openDelimiter = Delimiter.interpolationOpen;
  const closeDelimiter = Delimiter.interpolationClose;
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  const content = context.source.slice(openDelimiter.length, closeIndex);

  advanceBy(context, closeIndex + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context, length: number) {
  context.source = context.source.slice(length);
}

function createParseContext(content: string) {
  return {
    source: content,
  };
}
