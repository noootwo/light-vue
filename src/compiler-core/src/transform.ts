export const transform = (root, options?: any) => {
  const context = createTransformContext(root, options);
  traverseNode(root, context);

  createRootCodegen(root);
};

function createRootCodegen(root) {
  root.codegenNode = root.children[0];
}

function createTransformContext(root, options?) {
  return {
    root,
    nodeTransforms: options?.nodeTransforms || [],
  };
}

function traverseNode(root, context) {
  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    nodeTransforms[i](root);
  }

  if (root.children) {
    for (let i = 0; i < root.children.length; i++) {
      traverseNode(root.children[i], context);
    }
  }
}
