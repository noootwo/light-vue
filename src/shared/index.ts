export const EMPTY_OBJ = {};

export const extend = Object.assign;

export const isObject = (value: any) => {
  return value !== null && typeof value === "object";
};

export const hasChanged = (value, newValue) => !Object.is(value, newValue);

export const hasOwn = (val, key) => Object.hasOwnProperty.call(val, key);

export const camelize = (str: string) => {
  return str.replace(/-(\w)/g, (_, c: string) => {
    return c ? c.toUpperCase() : "";
  });
};

export const capitalize = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

export const toHandlerKey = (str: string) => {
  return str ? "on" + capitalize(str) : "";
};
