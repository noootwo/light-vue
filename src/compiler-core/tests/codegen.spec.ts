import { codegen } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";

describe("codegen", () => {
  it("happy path", () => {
    const ast = baseParse("hi");
    transform(ast);

    const code = codegen(ast);
    expect(code).toMatchSnapshot();
  });
});
