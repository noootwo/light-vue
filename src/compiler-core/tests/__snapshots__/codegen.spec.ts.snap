// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen happy path 1`] = `
Object {
  "code": "return function render(_ctx, _cache) {return 'hi'}",
}
`;
